app/src/gui/components/record/form.tsx

```diff
@@ -1032,6 +1059,99 @@ class RecordForm extends React.Component<RecordFormProps, RecordFormState> {
     );
   }
 
+  private handleCancel(ui_specification: ProjectUIModel) {
+    const relationState = this.props.location?.state;
+    console.log('cancelling', relationState, this.state.relationship);
+    // first case is if we have a parent record, there should be
+    // some location state passed in
+    if (relationState !== undefined && relationState !== null) {
+      // may need to undo a change to the parent record in the case
+      // that we are cancelling creation of a child record
+      if (relationState.parent_record_id && relationState.field_id) {
+        // need to edit the value of field_id in the parent
+        // to remove the reference to the child
+        this.handleCancelWithRelation({relationState, ui_specification});
+      }
+    } else {
+      const relationship = this.state.relationship;
+      if (
+        relationship === undefined ||
+        relationship === null ||
+        relationship.parent === undefined ||
+        relationship.parent === null
+      ) {
+        // there is no relationship, we're good to go
+        this.navigateTo(this.getRoute('project'));
+      } else {
+        generateLocationState(
+          relationship.parent,
+          this.props.project_id,
+          this.props.serverId
+        )
+          .then(locationState => {
+            return this.handleCancelWithRelation({
+              ui_specification,
+              relationState: locationState.location_state,
+            });
+          })
+          .catch(error => logError(error));
+      }
+    }
+    // remove the draft record
+    return deleteDraftsForRecord(this.props.project_id, this.props.record_id);
```

Comment:
Or does this always run - i.e. this.navigateTo will not actually prevent the execution of this code? The logic is a bit hard to follow but sensible as long as navigation does not prevent execution of the remainder of the code block

Reply:
Right, navigate doesn't exit the function, we carry on until here.